### PR TITLE
Optimize archive.ChangesDirs on Linux

### PR DESCRIPTION
**Pull request by burke**
**TL;DR: This makes `archive.Changesdirs` complete in ~20% of the previous runtime for mostly-unchanged layer comparisons on storage drivers where inodes line up between layers (incl. Overlay).**

We've been pained lately by the amount of time it takes docker to `commit` and `push` our images. After a bit of profiling, `ChangesDirs` seems to be the second worst offender (the largest being the `tarAppender`/`TarSum`/etc. pipeline, which I may take a look at too).

Our leaf images contain about 92,000 files, and take the current `archive.ChangesDirs` implementation about **3.2s** to chew through. With this change, that time is reduced to about **700ms**.

The major improvement here is that `getdents(2)` on linux actually returns inode numbers, but the Go stdlib throws them away very early on. The current implementation of `ChangesDirs` has to fan out to retrieve them one-by-one with `lstat(2)`. By walking both trees in parallel and ripping out the intermediate abstraction, propagating the inodes up from `getdents(2)`, we can prune identical files much earlier, saving a majority of `lstat(2)` calls.

I played with deferring the `lgetxattr(2)` calls until after comparing the data directly on the `stat_t` (ie. if the `mode` is different, no need to do anther syscall). It didn't make much of a difference in my workload and was marginally more complex, so I dropped it. I suspect it would give a more-than-marginal improvement if a large number of files had metadata modifications.

---
##### Other platforms

Some version of this should work in all UNIXes -- `readdir(3)` is a POSIX requirement and it's [specified to give inodes](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/dirent.h.html), but there's no standard syscall-level interface. It should be possible to do this abstractly by dropping to CGO and pulling in a libc for `readdir(3)`, and it should be safe to parse the output buffer the same across all UNIXes, but I'm not super jazzed on the idea of dropping to CGO. I'm also not keen on adding special-case code for, e.g., IllumOS or FreeBSD (both of which do appear to have at least roughly compatible `getdents(2)` signatures).

I don't know how to Windows.

### CODE REVIEW

**Comment by burke** [comment]
Heh, looks like rebasing onto master broke some things. Will fix...

**Comment by cpuguy83** [comment]
cc @unclejack

File: pkg/archive/changes_linux.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	if parent == nil {
+		return fmt.Errorf("collectFileInfoForChanges: Unexpectedly no parent for %s", path)
+	}
+	info := &FileInfo{
+		name:     filepath.Base(path),
+		children: make(map[string]*FileInfo),
+		parent:   parent,
+	}
+	cpath := filepath.Join(dir, path)
+	stat, err := system.FromStatT(fi.Sys().(*syscall.Stat_t))
```
Comment (burke) [inline pkg/archive/changes_linux.go]: The previous implementation actually calls `lstat(2)` twice on each file in both trees.

**Comment by unclejack** [comment]
@burke This looks good. I'm going to do some thorough testing. I'll review as well. Tests are passing, so that's a good first sign.

**Comment by thaJeztah** [comment]
@unclejack any update on this?

**Comment by icecrime** [comment]
Moving to code review.

**Comment by crosbymichael** [comment]
LGTM

**Comment by icecrime** [comment]
LGTM
